app/renderer/util.js

```diff
@@ -121,6 +137,119 @@ export function getOptimalXPath (doc, domNode, uniqueAttributes = ['id']) {
+          continue;
+        }
+
+        // If the attribute isn't actually unique, get it's index too
+        if (othersWithAttr.length > 1) {
+          let index = othersWithAttr.indexOf(domNode);
+          classChain = `${classChain}[${index + 1}]`;
+        }
+        return classChain;
+      }
+    }
+
+    // Get the relative xpath of this node using tagName
+    let classChain = `/${domNode.tagName}`;
+
+    // If this node has siblings of the same tagName, get the index of this node
+    if (domNode.parentNode) {
+      // Get the siblings
+      const childNodes = Array.prototype.slice.call(domNode.parentNode.childNodes, 0).filter((childNode) => (
+        childNode.nodeType === 1 && childNode.tagName === domNode.tagName
+      ));
+
+      // If there's more than one sibling, append the index
+      if (childNodes.length > 1) {
+        let index = childNodes.indexOf(domNode);
+        classChain += `[${index + 1}]`;
+      }
+    }
+
+    // Make a recursive call to this nodes parents and prepend it to this xpath
+    return getOptimalClassChain(doc, domNode.parentNode, uniqueAttributes) + classChain;
+  } catch (ign) {
+    // If there's an unexpected exception, abort and don't get an XPath
+    return null;
+  }
+}
+
+/**
+ * Get an optimal Predicate String for a DOMNode based on the getOptimalXPath method
+ * The `ios predicate string` can only search a single element, no parent child scope
+ *
+ * @param {DOMDocument} doc
+ * @param {DOMNode} domNode
+ * @param {Array<String>} uniqueAttributes Attributes we know are unique
+ * @returns {string}
+ */
+function getOptimalPredicateString (doc, domNode, uniqueAttributes) {
+  try {
+    // BASE CASE #1: If this isn't an element, we're above the root, or this is `XCUIElementTypeApplication`,
+    // which is not an official XCUITest element, return empty string
+    if (!domNode.tagName || domNode.nodeType !== 1 || domNode.tagName === 'XCUIElementTypeApplication') {
+      return '';
+    }
+
+    // BASE CASE #2: Check all attributes and try to find the best way
+    let xpathAttributes = [];
+    let predicateString = [];
+
+    for (let attrName of uniqueAttributes) {
+      const attrValue = domNode.getAttribute(attrName);
+      if (attrValue) {
+        xpathAttributes.push(`@${attrName}="${attrValue}"`);
+        let xpath = `//*[${xpathAttributes.join(' and ')}]`;
+        predicateString.push(`${attrName} == "${attrValue}"`);
+        let othersWithAttr;
+
+        // If the XPath does not parse, move to the next unique attribute
+        try {
+          othersWithAttr = XPath.select(xpath, doc);
+        } catch (ign) {
+          continue;
+        }
+
+        // If the attribute isn't actually unique, get it's index too
+        if (othersWithAttr.length === 1) {
+          return predicateString.join(' AND ');
+        }
+      }
+    }
+  } catch (ign) {
+    // If there's an unexpected exception, abort and don't get an XPath
+    return null;
```

Comment:
⬆️

Reply:
Think I fixed this

Reply:
@mykola-mokhnach 

Can you check if this is what you mean?

Reply:
I don't observe anything is done. I was expected to see something like `log.debug(e.stack)`

Reply:
Hmm I changed it to returning an empty string/. But the `return null` is exactly the same implementation as the original `getOptimalXPath` which I didn't change

Reply:
@mykola-mokhnach 

Added error log now, strange this is that you can't see it here, but you can see it in my latest commit.